### PR TITLE
Add rlimit values to thread_t

### PR DESCRIPTION
**Pull request by Sahakait**
We made the decision not to support setting rlimit values for the moment, as enforcing limits such file size would require significant changes without a current customer use-case to back up. 

Regardless, this PR preserves changes to rlimit setup that will allow support for setting rlimit values if needed in the future.

### CODE REVIEW

File: kernel/syscall.c [hunk 1]
```diff
@@ -2933,6 +2933,30 @@ long myst_syscall_prlimit64(
...
+        {
+            old_rlim->rlim_cur =  __myst_kernel_args.max_threads;
+            old_rlim->rlim_max =  __myst_kernel_args.max_threads;
+        }
+    }
+    else if (resource == RLIMIT_AS)
+    {
+        if (old_rlim)
+        {
+            old_rlim->rlim_cur =  -1;
```
Comment (vtikoo) [inline kernel/syscall.c:2948]: Is -1 a value with any special meaning here?

Comment (Sahakait) [reply kernel/syscall.c:2948]: Good call out, it means infinity and turns out there is a defined constant (RLIM_INFINITY).

File: kernel/syscall.c [hunk 2]
```diff
@@ -2933,6 +2933,30 @@ long myst_syscall_prlimit64(
...
     }
+    else if (resource == RLIMIT_NPROC)
+    {
+        if (old_rlim)
+        {
+            old_rlim->rlim_cur = __myst_kernel_args.max_threads;
+            old_rlim->rlim_max = __myst_kernel_args.max_threads;
+        }
+    }
+    else if (resource == RLIMIT_AS)
```
Comment (vtikoo) [inline kernel/syscall.c:2944]: The cases for RLIMIT_AS and RLIMIT_FSIZE can probably be folded together.

**Comment by vtikoo** [comment]
Could you also fix the reporting for RLIMIT_NOFILE in this PR? We are reporting an incorrect static value of 65536. It should be replaced with [MYST_FDTABLE_SIZE](https://github.com/deislabs/mystikos/blob/main/include/myst/fdtable.h#L22).

File: kernel/syscall.c (hunk 2 above)
Comment (Sahakait) [reply kernel/syscall.c:2944]: I would prefer to keep them as separate cases as we look into supporting setting RLIMIT.

File: kernel/syscall.c [hunk 3]
```diff
@@ -2933,6 +2933,30 @@ long myst_syscall_prlimit64(
...
+        }
+    }
+    else if (resource == RLIMIT_FSIZE)
+    {
+        if (old_rlim)
+        {
+            old_rlim->rlim_cur = RLIM_INFINITY;
+            old_rlim->rlim_max = RLIM_INFINITY;
+        }
+    }
```
Comment (bodzhang) [inline kernel/syscall.c:2959]: RLIMIT_DATA is relevant too. The app won't be able to allocate more than the Mystikos mman region for sure. I don't think it's a must-have for the implementation to enforce the memory allocation limit, but providing a "hint" on how much memory is available might allow some apps or runtimes to adjust their memory usage accordingly. Maybe the implementation can report a percentage of the mman region size, say 70%?

File: kernel/syscall.c [hunk 4]
```diff
@@ -2933,6 +2933,30 @@ long myst_syscall_prlimit64(
...
+        {
+            old_rlim->rlim_cur = __myst_kernel_args.max_threads;
+            old_rlim->rlim_max = __myst_kernel_args.max_threads;
+        }
+    }
+    else if (resource == RLIMIT_AS)
+    {
+        if (old_rlim)
+        {
+            old_rlim->rlim_cur = RLIM_INFINITY;
```
Comment (bodzhang) [inline kernel/syscall.c:2948]: App's address space can't exceed the Mystikos mman region. Maybe the Mystikos mman region size should be returned.

File: kernel/limit.c [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    if (resource < 0 || resource >= RLIM_NLIMITS)
+        ERAISE(-EINVAL);
+
+    if (resource == RLIMIT_MEMLOCK || resource == RLIMIT_SIGPENDING ||
+        resource == RLIMIT_MSGQUEUE)
+        ERAISE(-ENOTSUP);
+
+    if (pid == 0)
+        thread = myst_thread_self();
```
[the remark this answers is not in the document]
Comment (vtikoo) [reply kernel/limit.c:117]: `myst_find_process` is now available in `main`.

[the remark this answers is not in the document]
Comment (Sahakait) [reply kernel/limit.c:117]: Change has been made.

File: kernel/limit.c [hunk 6]
```diff
@@ -0,0 +1,148 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// #define _GNU_SOURCE
```
Comment (vtikoo) [inline kernel/limit.c:4]: Pls remove these headers if not needed

File: kernel/limit.c [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+    // RLIMIT_CPU (match linux)
+    rlimits[RLIMIT_CPU].rlim_cur = RLIM_INFINITY;
+    rlimits[RLIMIT_CPU].rlim_max = RLIM_INFINITY;
+
+    // RLIMIT_FSIZE (match linux)
+    rlimits[RLIMIT_FSIZE].rlim_cur = RLIM_INFINITY;
+    rlimits[RLIMIT_FSIZE].rlim_max = RLIM_INFINITY;
+
+    // RLIMIT_DATA (70% of mman region)
```
Comment (vtikoo) [inline kernel/limit.c:35]: Curious if theres any rationale behind this number?

File: kernel/syscall.c [hunk 8]
```diff
@@ -2909,41 +2909,15 @@ long myst_syscall_prlimit64(
...
-
-    if (resource == RLIMIT_NOFILE)
-    {
-        if (old_rlim)
-        {
-            // ATTN: return currently effective RLIMIT_NOFILE settings
-            old_rlim->rlim_cur = 65536;
-            old_rlim->rlim_max = 65536;
-        }
+    long ret = 0;
```
Comment (vtikoo) [inline kernel/syscall.c:2912]: this function should probably also be moved to limit.c.

File: tests/rlimit/rlimit.c [hunk 9]
```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
+
+// #include <myst/thread.h>
```
Comment (vtikoo) [inline tests/rlimit/rlimit.c:8]: remove commented lines

File: tests/rlimit/rlimit.c [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    struct rlimit rlim;
+    assert(getrlimit(resource, &rlim) == -1);
+    printf("%s: not supported \n", name);
+}
+
+int main(int argc, char* argv[])
+{
+    assert_rlimit(RLIMIT_CPU, "RLIMIT_CPU", RLIM_INFINITY, RLIM_INFINITY);
+    assert_rlimit(RLIMIT_FSIZE, "RLIMIT_FSIZE", RLIM_INFINITY, RLIM_INFINITY);
+    assert_rlimit_gt(RLIMIT_DATA, "RLIMIT_DATA", 40000000, 40000000);
```
Comment (vtikoo) [inline tests/rlimit/rlimit.c:58]: You could get the total ram via `SYS_sysinfo`.

File: kernel/limit.c (hunk 7 above)
Comment (Sahakait) [reply kernel/limit.c:35]: This was the suggested limit from Bo.

File: kernel/limit.c [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        ERAISE(-EFAULT);
+
+    if (pid < 0)
+        ERAISE(-EINVAL);
+
+    if (resource < 0 || resource >= RLIM_NLIMITS)
+        ERAISE(-EINVAL);
+
+    // ATTN: Setting rlimit value is currently unsupported
+    // Returning a failure will break tests/sysinfo and solutions/memcached
```
Comment (vtikoo) [inline kernel/limit.c:153]: Can you allow the silent passthrough only for `RLIMIT_NOFILE`? Thats the case which memcached requires.

Comment (Sahakait) [reply kernel/limit.c:153]: Done!

File: kernel/limit.c [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
```
Comment (vtikoo) [inline kernel/limit.c:153]: nit: tests/sysinfo no longer tests setrlimit.

**Review by vtikoo — APPROVED** [review]

**Pull request merged**